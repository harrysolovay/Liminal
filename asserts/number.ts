import { assert } from "../util/mod.ts"

export function min(value: number, min: number) {
  assert(value >= min, `Must be gte ${min}; received ${value}.`)
}

export function max(value: number, max: number) {
  assert(value <= max, `Must be lte ${max}; received ${value}.`)
}
