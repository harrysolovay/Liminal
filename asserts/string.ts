import { assert } from "../util/mod.ts"

export function minLength(value: string, minLength: number) {
  assert(value.length >= minLength, `Length must be gte ${minLength} chars.`)
}

export function maxLength(value: string, maxLength: number) {
  assert(value.length <= maxLength, `Length must be lte ${maxLength} chars.`)
}
