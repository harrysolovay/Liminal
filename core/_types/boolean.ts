import type { Type } from "../Type.ts"
import { declareType } from "./_declareType.ts"

export const boolean: Type<boolean> = declareType(() => boolean)
