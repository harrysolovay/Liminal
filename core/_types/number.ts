import type { Type } from "../Type.ts"
import { declareType } from "./_declareType.ts"

export const number: Type<number> = declareType(() => number)
