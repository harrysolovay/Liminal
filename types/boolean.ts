import { declare, type Type } from "../core/mod.ts"

export const boolean: Type.Initial<boolean> = declare({
  name: "boolean",
  source: {
    getType: () => boolean,
  },
  subschema: () => ({ type: "boolean" }),
})
