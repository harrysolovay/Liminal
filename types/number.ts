import { declareType, type Type } from "../core/mod.ts"

export const number: Type<number> = declareType({
  name: "number",
  source: {
    getType: () => number,
  },
})
