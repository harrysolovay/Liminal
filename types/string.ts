import { declare, type Type } from "../core/mod.ts"
import { assert } from "../util/assert.ts"

export const string: Type.Initial<string, {
  minLength: number
  maxLength: number
}> = declare({
  name: "string",
  source: {
    getType: () => string,
  },
  refinements: ({ minLength, maxLength }) => {
    assert(
      !(typeof minLength === "number" && typeof maxLength === "number") || minLength <= maxLength,
      "`minLength` cannot be greater than `maxLength`.",
    )
    return {
      minLength: `Length must be gte ${minLength} chars.`,
      maxLength: `Length must be lte ${maxLength} chars.`,
    }
  },
  subschema: () => ({ type: "string" }),
  output: (f) =>
    f({
      refinementPredicates: {
        minLength: (value, minLength) => value.length >= minLength,
        maxLength: (value, maxLength) => value.length <= maxLength,
      },
    }),
})
