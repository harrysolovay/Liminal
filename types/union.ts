import { type AnyType, declareType, type Type } from "../core/mod.ts"

export function union<M extends Array<AnyType>>(...members: M): Type<
  {
    [K in keyof M]: {
      type: K
      value: M[K]
    }
  }[number],
  M[number]["P"]
> {
  return declareType({
    name: "union",
    source: {
      factory: union,
      args: members,
    },
    visitValue: (value, visit) => visit(value.value, members[value.type]!),
  })
}
