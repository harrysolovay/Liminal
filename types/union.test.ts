import { assertTypeSnapshot } from "testing"
import { boolean } from "./boolean.ts"
import { number } from "./number.ts"
import { string } from "./string.ts"
import { union } from "./union.ts"

Deno.test("union", async (t) => {
  await assertTypeSnapshot(t, union(boolean, number, string))
})
