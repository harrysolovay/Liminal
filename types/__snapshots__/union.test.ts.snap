export const snapshot = {};

snapshot[`union 1`] = `
{
  anyOf: [
    {
      additionalProperties: false,
      properties: {
        type: {
          const: 0,
          type: "number",
        },
        value: {
          type: "boolean",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
    {
      additionalProperties: false,
      properties: {
        type: {
          const: 1,
          type: "number",
        },
        value: {
          type: "number",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
    {
      additionalProperties: false,
      properties: {
        type: {
          const: 2,
          type: "number",
        },
        value: {
          type: "string",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
  ],
  discriminator: "type",
}
`;

snapshot[`union 2`] = `
{
  anyOf: [
    {
      additionalProperties: false,
      properties: {
        type: {
          const: 0,
          type: "number",
        },
        value: {
          type: "boolean",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
    {
      additionalProperties: false,
      properties: {
        type: {
          const: 1,
          type: "number",
        },
        value: {
          type: "number",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
    {
      additionalProperties: false,
      properties: {
        type: {
          const: 2,
          type: "number",
        },
        value: {
          type: "string",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
  ],
  description: "One.",
  discriminator: "type",
}
`;

snapshot[`union 3`] = `
{
  anyOf: [
    {
      additionalProperties: false,
      properties: {
        type: {
          const: 0,
          type: "number",
        },
        value: {
          type: "boolean",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
    {
      additionalProperties: false,
      properties: {
        type: {
          const: 1,
          type: "number",
        },
        value: {
          type: "number",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
    {
      additionalProperties: false,
      properties: {
        type: {
          const: 2,
          type: "number",
        },
        value: {
          type: "string",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
  ],
  description: "One. Two.",
  discriminator: "type",
}
`;
